docs/design.md

```diff
@@ -0,0 +1,111 @@
+  [Hypothesis](http://hypothesis.works) for Ruby, that works with
+  RSpec (and ideally minitest, but if that at any point proves to
+  be a lot of work this may be dropped. It's not an explicit
+  requirement, but supporting it now makes it much easier to find
+  the right shape of the project design).
+* To provide a mostly feature complete version of the Conjecture
+  engine that powers Hypothesis in Rust, as decoupled from that
+  Ruby front-end as possible.
+
+Hypothesis for Ruby is not intended to be an exact feature for
+feature copy of the Python version. It will have a lot of the same
+underlying functionality, but with a number of changes driven by:
+
+* Trying to make it feel as "ruby native" as possible.
+* The ability to design an API from scratch that lacks many of the
+  constraints imposed both by the earlier much more limited functionality
+  of Hypothesis and the specifics of Python decorators and test
+  frameworks.
+
+## Differences
+
+The most fundamental API differences  between Hypothesis
+for Python and Hypothesis for Ruby are:
+
+* In Python we do a whole giant song and dance about exposing
+  functions for the test runner to call, while in Ruby we just
+  have a function which repeatedly calls a block and then fails.
+* in Python you specify a bunch of given parameters up front,
+  and then if you want values inline in the test you [explicitly
+  opt in to it](https://hypothesis.readthedocs.io/en/latest/data.html#drawing-interactively-in-tests),
+  while in Ruby this is not only the default but the only way to
+  get those values.
+* Strategies are called providers because strategy is a terrible
+  name that was originally intended to be internal and then leaked
+  into the public API because I wasn't thinking hard about naming.
+
+So for example:
+
+```ruby
+RSPec.describe "integer addition" do
+  it "commutes" do
+    hypothesis do
+      m = given integers
+      n = given integers
+      expect(m + n).to eq(n + m)
+    end
+  end
+end
+```
+
+```python
+@given(integers(), integers())
+def test_integers_commute(m, n):
+  assert m + n == n + m
+```
+
+The in-line style is slightly more verbose, but vastly more flexible
+and (I think) reads better. Also mixing in-line and up-front
+styles looks weird, and if we're going to have just one then
+the in-line approach is a strict superset of the functionality
+of the other.
+
+The main reason for these differences are:
+
+* Ruby blocks (and their relation to testing) make this approach
+  much more natural.
+* This functionality was not actually possible when the Hypothesis
+  for Python API was originally designed, which informed the way
+  its API looks.
+
+## Deliberate omissions
+
+The following are currently *not* part of the intended feature set
+of Hypothesis for Ruby:
+
+* Calls to `hypothesis` may not be nested.
+* None of [the functionality for adding explicit failures](https://hypothesis.readthedocs.io/en/latest/reproducing.html)
```

Comment:
I think we'll absolutely want explicit failures in the future (both with something `@example`-like and with an opaque representation of a problem test case). Doesn't seem necessary for the time-being, however.

Reply:
Yeah this is the omission I'm most sad about - `@example` is a very useful feature that is a large part of what distinguishes Hypothesis from other property-based testing libraries.

The reason for its omission is that this is currently not supported in the in-line style in Python either! I have a couple of ideas of how to do it, but I need to do a bunch of experimentation to figure out what a good API looks like and I don't want the delivery of this to be blocked on that.

If everything else gets done and there's enough time to spare, I will do that experimentation, but I think it would be optimistic to count on it getting in.